README.md
libfitbit
=========

by Kyle Machulis <kyle@nonpolynomial.com>

Nonpolynomial Labs - [http://www.nonpolynomial.com](http://www.nonpolynomial.com)

libfitbit is part of the OpenYou project for Open Source Health
Hardware access - [http://openyou.org](http://openyou.org)

If you find libfitbit useful, please donate to the project at
[http://pledgie.com/campaigns/14375](http://pledgie.com/campaigns/14375)


Credits and Thanks
------------------

Thanks to Matt Cutts for hooking me up with the hardware -
http://www.twitter.com/mattcutts


Description
-----------

libfitbit is an implementation of the data retrieval protocol for the
fitbit health tracking device. It also implements a synchronization
client for synchronizing data with the fitbit website on platforms not
supported by Fitbit currently.

The main goal of the library is to augment the services already
provided by the fitbit website, allowing for developers to make
interesting new applications that access the fitbit data in smaller
time chunks on the local machine, as well as giving users the ability
to back up their own data without having to rely on the web service.

Currently, the library only implements a proof of concept of the
protocol and web interface in python. This will be turned into a C
library, which can then be SWIG'd or otherwise interfaced to other
languages.

The fitbit reverse engineering document in the doc directory may also
be used for implementing the protocol in new languages without having
to read code (not that my ability to convey the protocol in english is
all that clear).


Package Information
-------------------

Source repo @ [http://www.github.com/qdot/libfitbit](http://www.github.com/qdot/libfitbit)


Platform Support
----------------

* Linux - Tested on Ubuntu 10.10
* OS X - Untested, should work?
* Windows - Won't work at the moment. May be able to create serial
  interface to talk to CP2012 chip? Haven't done research yet.


Library Requirements
--------------------

* Python - http://www.python.org
* libusb-1.0 - http://www.libusb.org
* pyusb 1.0+ - http://sourceforge.net/projects/pyusb/files/


Platform Cavaets
----------------

### Linux

You'll need to either run as root or set up a udev rule to switch out
permissions on the base VID/PID. We'll hopefully have a udev rule
checked in shortly.

### OS X

FitBit original driver is claiming the device resulting in premission errors
when libusb wants to claims it.

A solution that works everytime is simply to disable the driver by renaming
it:

```bash
cd /System/Library/Extensions
sudo mv SiLabsUSBDriver.kext SiLabsUSBDriver.kext.disabled
```

And reboot.

To re-enable it, just rename it again, and reboot again.

### Windows

Don't even know if it works there yet. :D


Future Plans
------------

* Breaking ANT access library out into its own repo
* Windows support
* Finish figuring out data types
* Implement library in C


License
-------

```
Copyright (c) 2011, Kyle Machulis/Nonpolynomial Labs
All rights reserved.

Redistribution and use in source and binary forms, 
with or without modification, are permitted provided 
that the following conditions are met:

   * Redistributions of source code must retain the 
     above copyright notice, this list of conditions 
     and the following disclaimer.
   * Redistributions in binary form must reproduce the 
     above copyright notice, this list of conditions and 
     the following disclaimer in the documentation and/or 
     other materials provided with the distribution.
   * Neither the name of the Nonpolynomial Labs nor the names 
     of its contributors may be used to endorse or promote 
     products derived from this software without specific 
     prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT 
NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, 
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```


ChangeLog.txt
libfitbit ChangeLog
=====================
Numbers refer to bug listings in the tracking system - https://github.com/openyou/libfitbit/issues

2011-02-07 Kyle Machulis <kyle@nonpolynomial.com>
	* Release v0.0.1
	* First version with working web sync, though not all steps run (can't burst data to device)


python/fitbit_client.py
#!/usr/bin/env python
#################################################################
# python fitbit web client for uploading data to fitbit site
# By Kyle Machulis <kyle@nonpolynomial.com>
# http://www.nonpolynomial.com
#
# Distributed as part of the libfitbit project
#
# Repo: http://www.github.com/openyou/libfitbit
#
# Licensed under the BSD License, as follows
#
# Copyright (c) 2011, Kyle Machulis/Nonpolynomial Labs
# All rights reserved.
#
# Redistribution and use in source and binary forms, 
# with or without modification, are permitted provided 
# that the following conditions are met:
#
#    * Redistributions of source code must retain the 
#      above copyright notice, this list of conditions 
#      and the following disclaimer.
#    * Redistributions in binary form must reproduce the 
#      above copyright notice, this list of conditions and 
#      the following disclaimer in the documentation and/or 
#      other materials provided with the distribution.
#    * Neither the name of the Nonpolynomial Labs nor the names 
#      of its contributors may be used to endorse or promote 
#      products derived from this software without specific 
#      prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
# CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT 
# NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, 
# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#################################################################

import sys
import urllib
import urllib2
import base64
import xml.etree.ElementTree as et
from fitbit import FitBit
from antprotocol.bases import FitBitANT, DynastreamANT

class FitBitResponse(object):
    def __init__(self, response):
        self.current_opcode = {}
        self.opcodes = []
        self.root = et.fromstring(response.strip())
        self.host = None
        self.path = None
        self.response = None
        if self.root.find("response") is not None:
            self.host = self.root.find("response").attrib["host"]
            self.path = self.root.find("response").attrib["path"]
            if self.root.find("response").text:
                # Quick and dirty url encode split
                self.response = dict([x.split("=") for x in urllib.unquote(self.root.find("response").text).split("&")])

        for opcode in self.root.findall("device/remoteOps/remoteOp"):
            op = {}
            op["opcode"] = [ord(x) for x in base64.b64decode(opcode.find("opCode").text)]
            op["payload"] = None
            if opcode.find("payloadData").text is not None:
                op["payload"] = [x for x in base64.b64decode(opcode.find("payloadData").text)]
            self.opcodes.append(op)

class FitBitClient(object):
    CLIENT_UUID = "2ea32002-a079-48f4-8020-0badd22939e3"
    FITBIT_HOST = "http://client.fitbit.com:80"
    START_PATH = "/device/tracker/uploadData"

    def __init__(self):
        self.info_dict = {}
        base = FitBitANT(debug=True)
        if not base.open():
            print "No devices connected!"
            return 1

        self.fitbit = FitBit(base)
        self.remote_info = None

    def form_base_info(self):
        self.info_dict.clear()
        self.info_dict["beaconType"] = "standard"
        self.info_dict["clientMode"] = "standard"
        self.info_dict["clientVersion"] = "1.3.3"
        self.info_dict["os"] = "Commodore 64"
        self.info_dict["clientId"] = self.CLIENT_UUID
        if self.remote_info:
            self.info_dict = dict(self.info_dict, **self.remote_info)

    def run_upload_request(self):
        self.fitbit.init_tracker_for_transfer()

        url = self.FITBIT_HOST + self.START_PATH

        # Start the request Chain
        self.form_base_info()
        while url is not None:
            res = urllib2.urlopen(url, urllib.urlencode(self.info_dict)).read()
            print res
            r = FitBitResponse(res)
            self.remote_info = r.response
            self.form_base_info()
            op_index = 0
            for o in r.opcodes:
                self.info_dict["opResponse[%d]" % op_index] = base64.b64encode(''.join([chr(x) for x in self.fitbit.run_opcode(o["opcode"], o["payload"])]))
                self.info_dict["opStatus[%d]" % op_index] = "success"
                op_index += 1
            urllib.urlencode(self.info_dict)
            print self.info_dict
            if r.host:
                url = "http://%s%s" % (r.host, r.path)
                print url
            else:
                print "No URL returned. Quitting."
                break
        self.fitbit.base.close()

def main():
    f = FitBitClient()
    f.run_upload_request()    
    return 0

if __name__ == '__main__':
    sys.exit(main())



python/antprotocol/bases.py
from .protocol import ANTReceiveException
from .libusb import ANTlibusb
import usb

class DynastreamANT(ANTlibusb):
    """Class that represents the Dynastream USB stick base, for
    garmin/suunto equipment. Only needs to set VID/PID.

    """
    VID = 0x0fcf
    PID = 0x1008

class FitBitANT(ANTlibusb):
    """Class that represents the fitbit base. Due to the extra
    hardware to handle tracker connection and charging, has an extra
    initialization sequence.

    """

    VID = 0x10c4
    PID = 0x84c4

    def open(self, vid = None, pid = None):
        if not super(FitBitANT, self).open(vid, pid):
            return False
        self.init()
        return True
    
    def init(self):
        # Device setup
        # bmRequestType, bmRequest, wValue, wIndex, data
        self._connection.ctrl_transfer(0x40, 0x00, 0xFFFF, 0x0, [])
        self._connection.ctrl_transfer(0x40, 0x01, 0x2000, 0x0, [])
        # At this point, we get a 4096 buffer, then start all over
        # again? Apparently doesn't require an explicit receive
        self._connection.ctrl_transfer(0x40, 0x00, 0x0, 0x0, [])
        self._connection.ctrl_transfer(0x40, 0x00, 0xFFFF, 0x0, [])
        self._connection.ctrl_transfer(0x40, 0x01, 0x2000, 0x0, [])
        self._connection.ctrl_transfer(0x40, 0x01, 0x4A, 0x0, [])
        # Receive 1 byte, should be 0x2
        self._connection.ctrl_transfer(0xC0, 0xFF, 0x370B, 0x0, 1)
        self._connection.ctrl_transfer(0x40, 0x03, 0x800, 0x0, [])
        self._connection.ctrl_transfer(0x40, 0x13, 0x0, 0x0, \
                                       [0x08, 0x00, 0x00, 0x00,
                                        0x40, 0x00, 0x00, 0x00,
                                        0x00, 0x00, 0x00, 0x00,
                                        0x00, 0x00, 0x00, 0x00
                                        ])
        self._connection.ctrl_transfer(0x40, 0x12, 0x0C, 0x0, [])
        try:
            self._receive()
        except usb.USBError:
            pass


python/antprotocol/libusb.py
#!/usr/bin/env python
#################################################################
# pyusb access for ant devices
# By Kyle Machulis <kyle@nonpolynomial.com>
# http://www.nonpolynomial.com
#
# Licensed under the BSD License, as follows
#
# Copyright (c) 2011, Kyle Machulis/Nonpolynomial Labs
# All rights reserved.
#
# Redistribution and use in source and binary forms, 
# with or without modification, are permitted provided 
# that the following conditions are met:
#
#    * Redistributions of source code must retain the 
#      above copyright notice, this list of conditions 
#      and the following disclaimer.
#    * Redistributions in binary form must reproduce the 
#      above copyright notice, this list of conditions and 
#      the following disclaimer in the documentation and/or 
#      other materials provided with the distribution.
#    * Neither the name of the Nonpolynomial Labs nor the names 
#      of its contributors may be used to endorse or promote 
#      products derived from this software without specific 
#      prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
# CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT 
# NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, 
# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#################################################################
#

from protocol import ANT, ANTReceiveException
import usb
import operator

class ANTlibusb(ANT):
    ep = { 'in'  : 0x81, \
           'out' : 0x01
           }

    def __init__(self, chan=0x0, debug=False):
        super(ANTlibusb, self).__init__(chan, debug)
        self._connection = False
        self.timeout = 1000

    def open(self, vid = None, pid = None):
        if vid is None:
            vid = self.VID
        if pid is None:
            pid = self.PID
        self._connection = usb.core.find(idVendor = vid,
                                         idProduct = pid)
        if self._connection is None:
            return False

        # For some reason, we have to set config, THEN reset,
        # otherwise we segfault back in the ctypes (on linux, at
        # least). 
        self._connection.set_configuration()
        self._connection.reset()
        # The we have to set our configuration again
        self._connection.set_configuration()

        # Then we should get back a reset check, with 0x80
        # (SUSPEND_RESET) as our status
        self._check_reset_response(0x80)
        return True

    def close(self):
        if self._connection is not None:
            self._connection = None

    def _send(self, command):
        # libusb expects ordinals, it'll redo the conversion itself.
        c = command
        self._connection.write(self.ep['out'], map(ord, c), 0, 100)

    def _receive(self, size=4096):
        r = self._connection.read(self.ep['in'], size, 0, self.timeout)
        if len(r) == 0:
            return r
        checksum = reduce(operator.xor, r[:-1])
        if checksum != r[-1]:
            raise ANTReceiveException("Checksums for packet do not match received values!")
        if self._debug:
            self.data_received(''.join(map(chr, r)))
        return r


python/fitbit.py
#!/usr/bin/env python
#################################################################
# python fitbit object
# By Kyle Machulis <kyle@nonpolynomial.com>
# http://www.nonpolynomial.com
#
# Distributed as part of the libfitbit project
#
# Repo: http://www.github.com/openyou/libfitbit
#
# Licensed under the BSD License, as follows
#
# Copyright (c) 2011, Kyle Machulis/Nonpolynomial Labs
# All rights reserved.
#
# Redistribution and use in source and binary forms,
# with or without modification, are permitted provided
# that the following conditions are met:
#
#    * Redistributions of source code must retain the
#      above copyright notice, this list of conditions
#      and the following disclaimer.
#    * Redistributions in binary form must reproduce the
#      above copyright notice, this list of conditions and
#      the following disclaimer in the documentation and/or
#      other materials provided with the distribution.
#    * Neither the name of the Nonpolynomial Labs nor the names
#      of its contributors may be used to endorse or promote
#      products derived from this software without specific
#      prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
# CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
# NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#################################################################
#
# ANT code originally taken from
# http://code.google.com/p/mstump-learning-exercises/source/browse/trunk/python/ANT/ant_twisted.py
# Added to and untwistedized and basically fixed up by Kyle Machulis <kyle@nonpolynomial.com>
#
# What's Done
#
# - Basic ANT protocol implementation
# - Basic FitBit protocol implementation
# - FitBit Base Initialization
# - FitBit Tracker Connection, Initialization, Info Retreival
# - Blind data retrieval (can get it, don't know what it is)
# - Talking to the fitbit website
# - Fix ANT Burst packet identifer
# - Add checksum checks for ANT receive
# - Fix packet status identifiers in ANT
#
# To Do (Big)
#
# - Dividing out into modules (ant classes may become their own library)
# - Figuring out more data formats and packets
# - Implementing data clearing

import itertools, sys, random, operator, datetime
from antprotocol.bases import FitBitANT, DynastreamANT
from antprotocol.protocol import ANTReceiveException

class FitBit(object):
    """Class to represent the fitbit tracker device, the portion of
    the fitbit worn by the user. Stores information about the tracker
    (serial number, firmware version, etc...).

    """

    def __init__(self, base = None):
        #: Iterator cycle of 0-8, for creating tracker packet serial numbers
        self.tracker_packet_count = itertools.cycle(range(0,8))

        # The tracker expects to start on 1, i.e. 0x39 This is set
        # after a reset (which is why we create the tracker in the
        # reset function). It won't talk if you try anything else.
        self.tracker_packet_count.next()

        #: used to track which internal databank we're on when
        self.current_bank_id = 0
        #: tracks current packet id for fitbit communication
        self.current_packet_id = None
        #: serial number of the tracker
        self.serial = None
        #: firmware version loaded on the tracker
        self.firmware_version = None
        #: Major version of BSL (?)
        self.bsl_major_version = None
        #: Minor version of BSL (?)
        self.bsl_minor_version = None
        #: Major version of App (?)
        self.app_major_version = None
        #: Minor version of App (?)
        self.app_minor_version = None
        #: True if tracker is in BSL Mode (?), False otherwise
        self.in_mode_bsl = None
        #: True if tracker is currently on charger, False otherwise
        self.on_charger = None

        self.base = base

    def gen_packet_id(self):
        """Generates the next packet id for information sent to the
        tracker.

        """

        self.current_packet_id = 0x38 + self.tracker_packet_count.next()
        return self.current_packet_id

    def parse_info_packet(self, data):
        """Parses the information gotten from the 0x24 retrieval command"""
        
        self.serial = data[0:5]
        self.firmware_version = data[5]
        self.bsl_major_version = data[6]
        self.bsl_minor_version = data[7]
        self.app_major_version = data[8]
        self.app_minor_version = data[9]
        self.in_mode_bsl = (False, True)[data[10]]
        self.on_charger = (False, True)[data[11]]

    def __str__(self):
        """Returns string representation of tracker information"""

        return "Tracker Serial: %s\n" \
               "Firmware Version: %d\n" \
               "BSL Version: %d.%d\n" \
               "APP Version: %d.%d\n" \
               "In Mode BSL? %s\n" \
               "On Charger? %s\n" % \
               ("".join(["%x" % (x) for x in self.serial]),
                self.firmware_version,
                self.bsl_major_version,
                self.bsl_minor_version,
                self.app_major_version,
                self.app_minor_version,
                self.in_mode_bsl,
                self.on_charger)

    def init_fitbit(self):
        self.init_device_channel([0xff, 0xff, 0x01, 0x01])

    def init_device_channel(self, channel):
        # ANT device initialization
        self.base.reset()
        self.base.send_network_key(0, [0,0,0,0,0,0,0,0])
        self.base.assign_channel()
        self.base.set_channel_period([0x0, 0x10])
        self.base.set_channel_frequency(0x2)
        self.base.set_transmit_power(0x3)
        self.base.set_search_timeout(0xFF)
        self.base.set_channel_id(channel)
        self.base.open_channel()

    def init_tracker_for_transfer(self):
        self.init_fitbit()
        self.wait_for_beacon()
        self.reset_tracker()

        # 0x78 0x02 is device id reset. This tells the device the new
        # channel id to hop to for dumpage
        cid = [random.randint(0,254), random.randint(0,254)]
        self.base.send_acknowledged_data([0x78, 0x02] + cid + [0x00, 0x00, 0x00, 0x00])
        self.base.close_channel()
        self.init_device_channel(cid + [0x01, 0x01])
        self.wait_for_beacon()
        self.ping_tracker()

    def reset_tracker(self):
        # 0x78 0x01 is apparently the device reset command
        self.base.send_acknowledged_data([0x78, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])

    def wait_for_beacon(self):
        # FitBit device initialization
        while 1:
            try:
                print "Waiting for receive"
                d = self.base._receive()
                if d[2] == 0x4E:
                    break
            except Exception:
                pass

    def _get_tracker_burst(self):
        d = self.base._check_burst_response()
        if d[1] != 0x81:
            raise Exception("Response received is not tracker burst! Got %s" % (d[0:2]))
        size = d[3] << 8 | d[2]
        if size == 0:
            return []
        return d[8:8+size]

    def run_opcode(self, opcode, payload = None):
        self.send_tracker_packet(opcode)
        data = self.base.receive_acknowledged_reply()
        if data[0] != self.current_packet_id:
            raise Exception("Tracker Packet IDs don't match! %02x %02x" % (data[0], self.current_packet_id))
        if data[1] == 0x42:
            return self.get_data_bank()
        if data[1] == 0x61:
            # Send payload data to device
            if payload is not None:
                self.send_tracker_payload(payload)
            data = self.base.receive_acknowledged_reply()
            if data[1] == 0x41:
                return [0x41, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]
        if data[1] == 0x41:
            return [0x41, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]

    def send_tracker_payload(self, payload):
        # The first packet will be the packet id, the length of the
        # payload, and ends with the payload CRC
        p = [0x00, self.gen_packet_id(), 0x80, len(payload), 0x00, 0x00, 0x00, 0x00, reduce(operator.xor, map(ord, payload))]
        prefix = itertools.cycle([0x20, 0x40, 0x60])
        for i in range(0, len(payload), 8):
            current_prefix = prefix.next()
            plist = []
            if i+8 >= len(payload):
                plist += [(current_prefix + 0x80) | self.base._chan]
            else:
                plist += [current_prefix | self.base._chan]
            plist += map(ord, payload[i:i+8])
            while len(plist) < 9:
                plist += [0x0]
            p += plist
        # TODO: Sending burst data with a guessed sleep value, should
        # probably be based on channel timing
        self.base._send_burst_data(p, .01)

    def get_tracker_info(self):
        data = self.run_opcode([0x24, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])
        self.parse_info_packet(data)
        return data

    def send_tracker_packet(self, packet):
        p = [self.gen_packet_id()] + packet
        self.base.send_acknowledged_data(p)

    def ping_tracker(self):
        self.base.send_acknowledged_data([0x78, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00])

    def check_tracker_data_bank(self, index):
        self.send_tracker_packet([0x60, 0x00, 0x02, index, 0x00, 0x00, 0x00])
        return self._get_tracker_burst()

    def run_data_bank_opcode(self, index):
        return self.run_opcode([0x22, index, 0x00, 0x00, 0x00, 0x00, 0x00])

    def get_data_bank(self):
        data = []
        while 1:
            try:
                bank = self.check_tracker_data_bank(self.current_bank_id)
                self.current_bank_id += 1
            except ANTReceiveException:
                continue
            if len(bank) == 0:
                return data
            data = data + bank

    def parse_bank2_data(self, data):
        for i in range(0, len(data), 13):
            print ["0x%.02x" % x for x in data[i:i+13]]
            # First 4 bytes are seconds from Jan 1, 1980
            print "Time: %s" % (datetime.datetime.fromtimestamp(data[i] | data[i + 1] << 8 | data[i + 2] << 16 | data[i + 3] << 24))

    def parse_bank0_data(self, data):
        # First 4 bytes are a time
        i = 0
        last_date_time = 0
        time_index = 0
        while i < len(data):
            # Date is in bigendian. No, really. And I think it's
            # because they're prefixing the 3 accelerometer reading
            # bytes with 0x80, so they can & against it.
            if not data[i] & 0x80:
                last_date_time = data[i+3] | data[i+2] << 8 | data[i+1] << 16 | data[i] << 24
                print "Time: %s" % (datetime.datetime.fromtimestamp(last_date_time))
                i = i + 4
                time_index = 0
            else:
                record_date = (datetime.datetime.fromtimestamp(last_date_time + 60 * time_index))
                # steps are easy. It's just the last byte
                steps = data[i+2] 
                # active score: second byte, subtract 10 (because METs
                # start at 1 but 1 is subtracted per minute, see
                # asterisk note on fitbit website, divide by 10.
                active_score = (data[i+1] - 10) / 10.0
                # first byte: I don't know. It starts at 0x81. So we at least subtract that.
                not_sure = data[i] - 0x81
                print "%s: ???: %d Active Score: %f Steps: %d" % (record_date, not_sure, active_score, steps)
                i = i + 3
                time_index = time_index + 1        

    def parse_bank1_data(self, data):
        for i in range(0, len(data), 14):
            print ["0x%.02x" % x for x in data[i:i+13]]
            # First 4 bytes are seconds from Jan 1, 1980
            daily_steps = data[i+7] << 8 | data[i+6]
            record_date = datetime.datetime.fromtimestamp(data[i] | data[i + 1] << 8 | data[i + 2] << 16 | data[i + 3] << 24)
            print "Time: %s Daily Steps: %d" % (record_date, daily_steps) 
        
def main():
    #base = DynastreamANT(True)
    base = FitBitANT(debug=True)
    if not base.open():
        print "No devices connected!"
        return 1

    device = FitBit(base)

    device.init_tracker_for_transfer()

    device.get_tracker_info()
    # print device.tracker

    device.parse_bank2_data(device.run_data_bank_opcode(0x02))
    print "---"
    device.parse_bank0_data(device.run_data_bank_opcode(0x00))
    device.run_data_bank_opcode(0x04)
    d = device.run_data_bank_opcode(0x02) # 13
    for i in range(0, len(d), 13):
        print ["%02x" % x for x in d[i:i+13]]
    d = device.run_data_bank_opcode(0x00) # 7
    print ["%02x" % x for x in d[0:7]]
    print ["%02x" % x for x in d[7:14]]
    j = 0
    for i in range(14, len(d), 3):
        print d[i:i+3]
        j += 1
    print "Records: %d" % (j)
    device.parse_bank1_data(device.run_data_bank_opcode(0x01))

    # for i in range(0, len(d), 14):
    #     print ["%02x" % x for x in d[i:i+14]]
    base.close()
    return 0

if __name__ == '__main__':
    sys.exit(main())
